src/hotspot/share/gc/z/zObjArrayAllocator.cpp

```diff
@@ -51,8 +56,20 @@ oop ZObjArrayAllocator::initialize(HeapWord* mem) const {
   // time and time-to-safepoint
   const size_t segment_max = ZUtils::bytes_to_words(64 * K);
 
-  if (_word_size <= segment_max || ArrayKlass::cast(_klass)->is_flatArray_klass()) {
-    // To small to use segmented clearing
+  if (_word_size <= segment_max) {
+    // Too small to use segmented clearing
+    return ObjArrayAllocator::initialize(mem);
+  }
+
+  if (is_oop_containing_flat_array(_klass)) {
```

Comment:
> A value object containing an oop and the added null-marker will always exceed 64 bits with ZGC, and therefore such objects will not be flattened in practice due to the 64-bit atomicity constraint.

Is this something we can guard against in the code as well? What happens if somebody creates a null-restricted non-atomic primitive-only flat array via the `ValueClass` API? Tearing from the user's POV is absolutely okay in such a case, but will any ZGC assumptions be violated?